### PR TITLE
WIP: Authentication

### LINKED ISSUE #23: Set up authentication backend
We need to authenticate users against our API.

We will have JWT based authentication and 3rd party authorization with GitHub and Twitter as alternatives.

Even if the users login with 3rd party providers they still need to be stored in the DB because we need to link their resources to an account so for such users we will only store email and provider.

For people who don't want to use third parties we will provide classical registration with email confirmation and JSON Web Token based authentication.

If at some point a user authenticates with 3rd party even though their email is already present in our db we will not create another user account, we will just add that provider to the existing account. So email should work as unique identifier for users. 

Users will be connected to networks and should preferably be stored in our OrientDB database.

### LINKED ISSUE #32: Create endpoint for lost account recovery
I just started poking around the auth implementation, aiming for a quick integration with the existing frontend.

So far I only explored the endpoints, looking for all the thing I might need in order to provide a full user experience. One issue I found is that there is no endpoint for recovering an account (e.g. "lost password" endpoint).

The desired user story would be:

* user has an unauthenticated endpoint where she can `POST` something like this: `{ "email": "contact@emxample.org" }`
* server registers the request and generates a one time password refresh token
* user receives an email with the password refresh token
  * user `POST`s back the password refresh token together with the new password
* server processes this as:
  * validates token
  * expires the token
  * changes user password
  * returns HTTP 200 for the request

### PR DESCRIPTION
**Pull request by georgiana-b**
Fixes #23 
Fixes #32

### CODE REVIEW

File: test/api/auth.js [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+});
+
+test.serial('login: Success', async (t) => {
+  t.plan(4);
+
+  const res = await request(app)
+    .post(LOGIN_ROUTE)
+    .send({ email: 'testemail123456@mailinator.com', password: '123456789test' });
+
+  refreshToken = res.body.data.refreshToken;
```
[the remark this answers is not in the document]
Comment (edgarmarkosov) [reply test/api/auth.js]: I am using it later for checking if refreshing the token is working with using refresh token.

File: test/api/auth.js [hunk 2]
```diff
@@ -70,21 +77,33 @@ test.serial('register: Email is taken', async (t) => {
...
-  refreshToken = res.body.data.refreshToken;
   t.is(res.status, SUCCESS);
   t.truthy(res.body.data);
   t.truthy(res.body.data.accessToken);
   t.truthy(res.body.data.refreshToken);
+
+  t.log('refresh:token: Success');
+
+  const res1 = await request(app)
+    .post(REFRESH_TOKEN_ROUTE)
```
Comment (georgiana-b) [inline test/api/auth.js]: Wouldn't it be better if this was a separate test?

File: test/api/auth.js [hunk 3]
```diff
@@ -109,16 +128,26 @@ test.serial('login: Wrong password', async (t) => {
...
-test.serial('refresh:token: Success', async (t) => {
-  t.plan(4);
-
-  const res = await request(app)
-    .post(REFRESH_TOKEN_ROUTE)
-    .set('x-refresh-token', refreshToken)
-    .send({});
+test.cb('passport:github', (t) => {
+  t.plan(5);
+  passport.githubStrategyCallback(null, null, { id: '123456', emails: [{ value: 'testemail123456@mailinator.com', primary: true }] }, (err, user) => {
```
Comment (georgiana-b) [inline test/api/auth.js]: Neat!

File: test/api/auth.js (hunk 2 above)
Comment (edgarmarkosov) [reply test/api/auth.js]: I need logged in user in order to use his refresh token. Db is truncated after each test, that is why I merged them in 1 test.

File: test/api/auth.js (hunk 3 above)
Comment (edgarmarkosov) [reply test/api/auth.js]: Thanks :)

**Comment by georgiana-b** [comment]
Opened #41 for refresh token tests.

**Pull request merged**